File: example/python/vanilla_swap.py
"""Build and price a vanilla fixed-vs-floating interest-rate swap.

Two ways to get the same swap:

* `MakeVanillaSwap` - a builder that fills in the market conventions (fixed-leg
  tenor and day count, floating leg from the index) from a tenor and an index.
* `VanillaSwap` - the explicit form, where every schedule and day count is
  spelled out by hand.

The swap is priced by a `DiscountingSwapEngine`, which here is attached with
`swap.set_engine(curve, settings)` (the engine discounts both legs off the given
curve). The fixture is QuantLib's Jamshidian 5Y swap; the fair rate and NPV
below reproduce the pins in `test_vanilla_swap.py`, so they double as a
self-check.

Run it with:

    python example/python/vanilla_swap.py
"""

# plugins
# itofin library
from itofin import Settings
from itofin.indexes import Euribor
from itofin.instruments import MakeVanillaSwap, SwapType, VanillaSwap
from itofin.termstructures import FlatForward
from itofin.time import BusinessDayConvention, Calendar, Date, DayCounter, Frequency, Period, Schedule

REF = Date(15, 1, 2026)
START = Date(15, 1, 2028)
END = Date(15, 1, 2033)


def make_market():
    settings = Settings()
    settings.set_evaluation_date(REF)
    # A flat 3% continuous curve for both forwarding and discounting.
    curve = FlatForward(REF, 0.03, DayCounter.actual365_fixed())
    index = Euribor.six_months(curve, settings)
    return settings, curve, index


def priced_with_builder():
    """A par 5Y payer swap via the builder: `fixed_rate=None` fills the fixed
    leg with the fair rate, so the swap prices to zero."""
    settings, _curve, index = make_market()
    swap = MakeVanillaSwap(
        Period(5, "Years"),
        index,
        settings,
        fixed_rate=None,
        effective_date=START,
        nominal=100.0,
    ).build()  # MakeVanillaSwap is a builder: .build() returns the swap.
    return swap


def priced_by_hand():
    """The same swap spelled out: annual fixed leg on 30/360 bond basis,
    semiannual floating leg on Euribor6M / Actual360, priced at a 3% coupon."""
    settings, curve, index = make_market()
    fixed = Schedule(
        START,
        END,
        Frequency.Annual,
        Calendar.target(),
        BusinessDayConvention.ModifiedFollowing,
    )
    floating = Schedule(
        START,
        END,
        Frequency.Semiannual,
        Calendar.target(),
        BusinessDayConvention.ModifiedFollowing,
    )
    swap = VanillaSwap(
        SwapType.Payer,
        100.0,  # nominal
        fixed,
        0.03,  # fixed rate
        DayCounter.thirty360_bond_basis(),
        floating,
        index,
        0.0,  # floating spread
        DayCounter.actual360(),
        settings,
    )
    # DiscountingSwapEngine off `curve`: VanillaSwap.set_engine takes the curve
    # and settings directly, not a separate engine object.
    swap.set_engine(curve, settings)
    return swap


def main() -> None:
    par = priced_with_builder()
    print("MakeVanillaSwap, par 5Y payer (fixed_rate=None):")
    print(f"  fair_rate = {par.fair_rate() * 100:.6f}%")
    print(f"  fixed_rate(filled) = {par.fixed_rate() * 100:.6f}%")
    print(f"  NPV       = {par.npv():.10f}  (par swap -> ~0)")

    hand = priced_by_hand()
    print("\nHand-built VanillaSwap, 5Y payer at a 3% coupon:")
    print(f"  nominal   = {hand.nominal():.2f}")
    print(f"  fixed_rate= {hand.fixed_rate() * 100:.4f}%")
    print(f"  fair_rate = {hand.fair_rate() * 100:.6f}%")
    print(f"  NPV       = {hand.npv():.10f}")


if __name__ == "__main__":
    main()

File: crates/libitofin/examples/vanilla_swap.rs
//! Vanilla fixed-vs-float interest-rate swap priced with `DiscountingSwapEngine`.
//!
//! Built through `MakeVanillaSwap` (the QuantLib `MakeVanillaSwap` port), which
//! derives both schedules, calls `VanillaSwap::new`, and attaches a
//! `DiscountingSwapEngine` over the index's forwarding curve. A 5Y EUR swap:
//! annual Thirty360(BondBasis) fixed leg (EUR currency default) vs semiannual
//! Euribor 6M / Actual360 floating leg, both forecasting and discounting off a
//! flat 2% continuously-compounded curve.
//!
//! D5: `Settings` is an explicit `Shared` handle carrying the evaluation date
//! (no global singleton). D11: because the swap starts on a future effective
//! date, every floating fixing lies in the future and is forecast off the curve,
//! so no past fixing needs to be seeded into `Settings`' fixing store.

use libitofin::handle::Handle;
use libitofin::indexes::IborIndex;
use libitofin::indexes::ibor::Euribor;
use libitofin::instrument::Instrument; // brings the `npv()` method into scope
use libitofin::instruments::MakeVanillaSwap;
use libitofin::interestrate::Compounding;
use libitofin::settings::Settings;
use libitofin::shared::{Shared, shared};
use libitofin::termstructures::yields::FlatForward;
use libitofin::termstructures::yieldtermstructure::YieldTermStructure;
use libitofin::time::date::{Date, Month};
use libitofin::time::daycounters::actual360::Actual360;
use libitofin::time::frequency::Frequency;
use libitofin::time::period::Period;
use libitofin::time::timeunit::TimeUnit;

fn main() -> libitofin::errors::QlResult<()> {
    // --- D5: explicit Settings with an evaluation date (no global clock) ---
    let today = Date::new(7, Month::July, 2026);
    let settings: Shared<Settings<Date>> = shared(Settings::<Date>::new());
    settings.set_evaluation_date(today);

    // --- Flat 2% forecasting/discounting curve, wrapped in a Handle ---
    let curve: Handle<dyn YieldTermStructure> = Handle::new(shared(FlatForward::with_rate(
        today,
        0.02,
        Actual360::new(),
        Compounding::Continuous,
        Frequency::Annual,
    ))
        as Shared<dyn YieldTermStructure>);

    // --- 6-month Euribor forecasting off that curve ---
    let index: Shared<IborIndex> = shared(Euribor::six_months(curve, Shared::clone(&settings)));

    // --- Build + price a 5Y payer swap at a fixed 3% via MakeVanillaSwap ---
    // Effective date after the evaluation date => all fixings are forecast
    // (no stored past fixing required). build() attaches DiscountingSwapEngine.
    let mut swap = MakeVanillaSwap::new(
        Period::new(5, TimeUnit::Years),
        Shared::clone(&index),
        Some(0.03),                     // fixed rate; pass None to fill the fair rate
        Period::new(0, TimeUnit::Days), // no forward start
        Shared::clone(&settings),
    )
    .with_effective_date(Date::new(9, Month::July, 2026))
    .build()?;

    // npv() (Instrument trait) prices via the attached DiscountingSwapEngine.
    let npv = swap.npv()?;
    // fair (par) swap rate is on the FixedVsFloatingSwap base, needs &mut.
    let fair_rate = swap.fixed_vs_floating_mut().fair_rate()?;

    println!("Swap NPV:  {npv:.6}");
    println!("Fair rate: {:.6}%", fair_rate * 100.0);
    Ok(())
}
